### PR TITLE
runtime: Parse integers as booleans when necessary

### PR DESCRIPTION
**Pull request by tonya11en**
This patch adds the ability for runtime booleans to also parse `0` and `1` as false and true respectively. This removes certain classes of operator configuration mistakes.

Risk Level: Low
Testing: Unit test
Docs Changes: n/a
Release Notes: done

### CODE REVIEW

File: source/common/runtime/runtime_impl.cc [hunk 1]
```diff
@@ -342,10 +342,10 @@ bool SnapshotImpl::parseEntryBooleanValue(Entry& entry) {
   absl::string_view stripped = entry.raw_string_value_;
   stripped = absl::StripAsciiWhitespace(stripped);
 
-  if (absl::EqualsIgnoreCase(stripped, "true")) {
+  if (absl::EqualsIgnoreCase(stripped, "true") || absl::EqualsIgnoreCase(stripped, "1")) {
```
Comment (mattklein123) [inline source/common/runtime/runtime_impl.cc]: I could go either way on this, but what about trying to parse as an int, and if an init, 0 is false and anything other than 0 is true?

**Comment by mattklein123** [comment]
The other thing I think we can do here: Every time we load a new runtime layer, I think we could enumerate all of the feature enable/disable flags and warn/inc failure stat if any of them are not bools. This would make it much more clear that there was operator error. I don't think this would be too hard to do? cc @alyssawilk

File: source/common/runtime/runtime_impl.cc (hunk 1 above)
Comment (tonya11en) [reply source/common/runtime/runtime_impl.cc]: That's a much better idea.

File: docs/root/configuration/operations/runtime.rst [hunk 2]
```diff
@@ -255,6 +255,14 @@ with the new code paths are flushed out ahead of time, rather than after the cod
 
 .. _runtime_stats:
 
+.. attention::
+
+   Versions of Envoy prior to 1.14.1 cannot parse runtime booleans from integer values and require
+   an explicit "true" or "false". Mistakenly placing an integer such as "0" to represent "false"
+   will lead to usage of the default value. This is especially important to keep in mind for case of
+   runtime overrides for deprecated features, as it will can potentially result in unexpected Envoy
```
Comment (mattklein123) [inline docs/root/configuration/operations/runtime.rst]: Can you ref link to the deprecated docs?

File: source/common/runtime/runtime_impl.cc [hunk 3]
```diff
@@ -342,7 +342,13 @@ bool SnapshotImpl::parseEntryBooleanValue(Entry& entry) {
   absl::string_view stripped = entry.raw_string_value_;
   stripped = absl::StripAsciiWhitespace(stripped);
 
-  if (absl::EqualsIgnoreCase(stripped, "true")) {
+  uint64_t parse_int;
+  if (absl::SimpleAtoi(stripped, &parse_int)) {
+    entry.bool_value_ = static_cast<bool>(parse_int);
```
Comment (mattklein123) [inline source/common/runtime/runtime_impl.cc]: nit: prefer `parse_int != 0`

File: test/common/runtime/runtime_impl_test.cc [hunk 4]
```diff
@@ -653,6 +655,17 @@ TEST_F(StaticLoaderImplTest, ProtoParsing) {
   EXPECT_EQ(false, snapshot->getBoolean("file13", true));
   EXPECT_EQ(false, snapshot->getBoolean("file13", false));
 
+  EXPECT_EQ(false, snapshot->getBoolean("bool_as_int0", true));
+  EXPECT_EQ(false, snapshot->getBoolean("bool_as_int0", false));
+  EXPECT_EQ(true, snapshot->getBoolean("bool_as_int1", false));
+  EXPECT_EQ(true, snapshot->getBoolean("bool_as_int1", true));
```
Comment (mattklein123) [inline test/common/runtime/runtime_impl_test.cc]: Please test that this also parsed as an int.

**Pull request merged**